actionview/lib/action_view/helpers/text_helper.rb

```diff
@@ -139,16 +139,19 @@ def highlight(text, phrases, options = {}, &block)
         if text.blank? || phrases.blank?
           text || ""
         else
-          match = Array(phrases).map do |p|
-            Regexp === p ? p.to_s : Regexp.escape(p)
-          end.join("|")
-
-          if block_given?
-            text.gsub(/(#{match})(?![^<]*?>)/i, &block)
-          else
-            highlighter = options.fetch(:highlighter, '<mark>\1</mark>')
-            text.gsub(/(#{match})(?![^<]*?>)/i, highlighter)
-          end
+          patterns = Array(phrases).map { |phrase| Regexp === phrase ? phrase : Regexp.escape(phrase) }
+          pattern = /(#{patterns.join("|")})/i
+          highlighter = options.fetch(:highlighter, '<mark>\1</mark>') unless block
+
+          text.scan(/<[^>]+>|[^<]+/).each do |segment|
```

Comment:
This regex has backtracking issues on Ruby < 3.2 for example with the string `"<"*99999`. I think that's only possible if this is called with `sanitize: false`, which seems like a misuse of this API (and actually this change alters that behaviour too), but maybe we could find an alternative way to do this?

Reply:
Interesting point!  I think the following works as well:

```suggestion
          text.scan(/<[^>]*|[^<]+/).each do |segment|
```

And doesn't exhibit backtracking issues:

```ruby
string = "<"*99999

puts Benchmark.measure { string.scan(/<[^>]+>|[^<]+/){} }
# 91.561392   0.019090  91.580482 ( 92.037754)

puts Benchmark.measure { string.scan(/<[^>]*|[^<]+/){} }
# 0.003507   0.000000   0.003507 (  0.005074)
```

I actually like that regex better, because it should ensure that no portion of the string is dropped by the `scan`, even if the string includes malformed HTML.  (Though, some highlighting might still be skipped in that case, but I think that's a reasonable trade-off.)

Reply:
Applied this change.